File: problems/671-second-minimum-node-in-a-binary-tree/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int findSecondMinimumValue(struct TreeNode* root) {
    long min1 = (long)root->val;
    long min2 = (long)INT_MAX;
    
    if(root->left){
        long leftVal = (long)root->left->val;
        if(leftVal > min1){
            min2 = min(min2,leftVal);
        } else {
            min2 = min(min2,findSecondMinimumValue(root->left));
        }
    }
    if(root->right){
        long rightVal = (long)root->right->val;
        if(rightVal > min1){
            min2 = min(min2,rightVal);
        } else {
            min2 = min(min2,findSecondMinimumValue(root->right));
        }
    }
    
    if(min2 == (long)INT_MAX) return -1;
    return (int)min2;
}

File: problems/671-second-minimum-node-in-a-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int findSecondMinimumValue(TreeNode root) {
        long min1 = Long.MAX_VALUE;
        long min2 = Long.MAX_VALUE;

        dfs(root, min1, min2);

        return min2 == Long.MAX_VALUE ? -1 : (int)min2;
    }


    private void dfs(TreeNode node, long min1, long min2){
        if(node == null) return;

        if(node.val < min1){
            min2 = min1;
            min1 = node.val;
        } else if (node.val < min2 && node.val > min1){
            min2 = node.val;
        }

        dfs(node.left, min1, min2);
        dfs(node.right, min1, min2);
    }
}